File: scripts/compress-image.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import sharp from "sharp";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");

const DEFAULT_SIZE = 400;
const DEFAULT_QUALITY = 70;
const IMAGE_EXT = new Set([".png", ".webp", ".jpg", ".jpeg"]);

function printHelp() {
  console.log(`Usage:
  pnpm compress-image <input> <output.webp> [options]
  pnpm compress-image -- --batch <input-dir> <output-dir> [options]

Options:
  --size <px>       Output canvas size (default: ${DEFAULT_SIZE})
  --quality <n>     WebP quality 1-100 (default: ${DEFAULT_QUALITY})
  --in-place        Overwrite input file (output path optional)

Examples:
  pnpm compress-image public/images/msc/about.webp public/images/msc/about.webp --in-place
  pnpm compress-image sprite.webp out.webp --quality 70 --size 400
  pnpm compress-image -- --batch public/images/gods public/images/gods --in-place
`);
}

function parseArgs(argv) {
  const options = {
    batch: null,
    inPlace: false,
    size: DEFAULT_SIZE,
    quality: DEFAULT_QUALITY,
    positional: [],
  };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--batch") {
      options.batch = [argv[i + 1], argv[i + 2]];
      i += 2;
      continue;
    }
    if (arg === "--in-place") {
      options.inPlace = true;
      continue;
    }
    if (arg === "--size") {
      options.size = Number(argv[i + 1]);
      i += 1;
      continue;
    }
    if (arg === "--quality") {
      options.quality = Number(argv[i + 1]);
      i += 1;
      continue;
    }
    if (arg.startsWith("-")) {
      console.error(`[compress-image] Unknown option: ${arg}`);
      process.exit(1);
    }
    options.positional.push(arg);
  }

  return options;
}

function formatBytes(bytes) {
  if (bytes < 1024) {
    return `${bytes} B`;
  }
  return `${(bytes / 1024).toFixed(1)} KB`;
}

async function compressFile(inputPath, outputPath, { size, quality }) {
  const inputStat = fs.statSync(inputPath);
  const inputBuffer = fs.readFileSync(inputPath);

  await sharp(inputBuffer)
    .resize(size, size, {
      fit: "contain",
      background: { r: 0, g: 0, b: 0, alpha: 0 },
      kernel: sharp.kernel.nearest,
    })
    .webp({ quality, effort: 6 })
    .toFile(outputPath);

  const outputStat = fs.statSync(outputPath);
  const meta = await sharp(outputPath).metadata();
  console.log(
    `${path.relative(root, outputPath)}: ${meta.width}x${meta.height} q=${quality} ` +
      `${formatBytes(inputStat.size)} → ${formatBytes(outputStat.size)}`
  );
}

function listImages(dir) {
  const files = [];
  const walk = (current) => {
    for (const entry of fs.readdirSync(current, { withFileTypes: true })) {
      const full = path.join(current, entry.name);
      if (entry.isDirectory()) {
        walk(full);
        continue;
      }
      if (IMAGE_EXT.has(path.extname(entry.name).toLowerCase())) {
        files.push(full);
      }
    }
  };
  walk(dir);
  return files.sort();
}

async function main() {
  const argv = process.argv.slice(2);
  if (argv.length === 0 || argv.includes("-h") || argv.includes("--help")) {
    printHelp();
    return;
  }

  const options = parseArgs(argv);

  if (options.quality < 1 || options.quality > 100) {
    console.error("[compress-image] --quality must be between 1 and 100");
    process.exit(1);
  }

  if (options.size < 1) {
    console.error("[compress-image] --size must be a positive integer");
    process.exit(1);
  }

  if (options.batch) {
    const [inputDir, outputDir] = options.batch.map((p) => path.resolve(p));
    if (!fs.existsSync(inputDir)) {
      console.error(`[compress-image] Directory not found: ${inputDir}`);
      process.exit(1);
    }

    const files = listImages(inputDir);
    if (files.length === 0) {
      console.error(`[compress-image] No images in ${inputDir}`);
      process.exit(1);
    }

    for (const inputPath of files) {
      const rel = path.relative(inputDir, inputPath);
      const outputPath = options.inPlace
        ? inputPath
        : path.join(outputDir, rel).replace(/\.(png|jpe?g)$/i, ".webp");

      if (!options.inPlace) {
        fs.mkdirSync(path.dirname(outputPath), { recursive: true });
      }

      const tempPath = options.inPlace
        ? `${inputPath}.tmp.webp`
        : outputPath;

      await compressFile(inputPath, tempPath, options);

      if (options.inPlace) {
        fs.renameSync(tempPath, inputPath);
      }
    }
    return;
  }

  const [inputArg, outputArg] = options.positional;
  if (!inputArg) {
    printHelp();
    process.exit(1);
  }

  const inputPath = path.resolve(inputArg);
  if (!fs.existsSync(inputPath)) {
    console.error(`[compress-image] File not found: ${inputPath}`);
    process.exit(1);
  }

  const outputPath = options.inPlace
    ? inputPath
    : path.resolve(outputArg ?? inputPath);

  if (!options.inPlace && !outputArg) {
    console.error("[compress-image] Output path required unless --in-place is set");
    process.exit(1);
  }

  const tempPath = options.inPlace ? `${inputPath}.tmp.webp` : outputPath;
  fs.mkdirSync(path.dirname(tempPath), { recursive: true });

  await compressFile(inputPath, tempPath, options);

  if (options.inPlace) {
    fs.renameSync(tempPath, inputPath);
  }
}

main().catch((error) => {
  console.error(`[compress-image] ${error.message}`);
  process.exit(1);
});

File: scripts/compress-image.py
#!/usr/bin/env python3
"""Resize and compress game sprites to WebP (nearest-neighbor, transparent canvas)."""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

_SCRIPT_DIR = Path(__file__).resolve().parent
if str(_SCRIPT_DIR) not in sys.path:
    sys.path.insert(0, str(_SCRIPT_DIR))

from deps import ensure_imports

ensure_imports("PIL")

from PIL import Image

ROOT = _SCRIPT_DIR.parent
DEFAULT_SIZE = 400
DEFAULT_QUALITY = 70
IMAGE_EXT = {".png", ".webp", ".jpg", ".jpeg"}


def format_bytes(size: int) -> str:
    if size < 1024:
        return f"{size} B"
    return f"{size / 1024:.1f} KB"


def resize_contain(img: Image.Image, size: int) -> Image.Image:
    rgba = img.convert("RGBA")
    width, height = rgba.size
    scale = min(size / width, size / height)
    new_w = max(1, int(width * scale))
    new_h = max(1, int(height * scale))
    resized = rgba.resize((new_w, new_h), Image.Resampling.NEAREST)
    canvas = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    offset = ((size - new_w) // 2, (size - new_h) // 2)
    canvas.paste(resized, offset)
    return canvas


def compress_file(input_path: Path, output_path: Path, *, size: int, quality: int) -> None:
    input_stat = input_path.stat()
    with Image.open(input_path) as img:
        out = resize_contain(img, size)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        out.save(output_path, format="WEBP", quality=quality, method=6)

    output_stat = output_path.stat()
    with Image.open(output_path) as saved:
        width, height = saved.size
    rel = output_path.relative_to(ROOT)
    print(
        f"{rel}: {width}x{height} q={quality} "
        f"{format_bytes(input_stat.st_size)} → {format_bytes(output_stat.st_size)}"
    )


def list_images(directory: Path) -> list[Path]:
    files: list[Path] = []

    def walk(current: Path) -> None:
        for entry in sorted(current.iterdir()):
            if entry.is_dir():
                walk(entry)
            elif entry.suffix.lower() in IMAGE_EXT:
                files.append(entry)

    walk(directory)
    return files


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("input", nargs="?", help="Input image path")
    parser.add_argument("output", nargs="?", help="Output .webp path")
    parser.add_argument("--batch", nargs=2, metavar=("INPUT_DIR", "OUTPUT_DIR"))
    parser.add_argument("--in-place", action="store_true")
    parser.add_argument("--size", type=int, default=DEFAULT_SIZE)
    parser.add_argument("--quality", type=int, default=DEFAULT_QUALITY)
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv or sys.argv[1:])

    if args.quality < 1 or args.quality > 100:
        print("[compress-image] --quality must be between 1 and 100", file=sys.stderr)
        return 1

    if args.size < 1:
        print("[compress-image] --size must be a positive integer", file=sys.stderr)
        return 1

    if args.batch:
        input_dir = Path(args.batch[0]).resolve()
        output_dir = Path(args.batch[1]).resolve()
        if not input_dir.is_dir():
            print(f"[compress-image] Directory not found: {input_dir}", file=sys.stderr)
            return 1

        files = list_images(input_dir)
        if not files:
            print(f"[compress-image] No images in {input_dir}", file=sys.stderr)
            return 1

        for input_path in files:
            rel = input_path.relative_to(input_dir)
            if args.in_place:
                output_path = input_path
                temp_path = input_path.with_suffix(f"{input_path.suffix}.tmp.webp")
            else:
                output_path = (output_dir / rel).with_suffix(".webp")
                temp_path = output_path

            compress_file(input_path, temp_path, size=args.size, quality=args.quality)
            if args.in_place:
                temp_path.replace(input_path)
        return 0

    if not args.input:
        print(
            "usage: compress-image.py <input> <output.webp> [options]\n"
            "       compress-image.py --batch <input-dir> <output-dir> [options]",
            file=sys.stderr,
        )
        return 1

    input_path = Path(args.input).resolve()
    if not input_path.is_file():
        print(f"[compress-image] File not found: {input_path}", file=sys.stderr)
        return 1

    if args.in_place:
        output_path = input_path
        temp_path = input_path.with_suffix(f"{input_path.suffix}.tmp.webp")
    else:
        if not args.output:
            print(
                "[compress-image] Output path required unless --in-place is set",
                file=sys.stderr,
            )
            return 1
        output_path = Path(args.output).resolve()
        temp_path = output_path

    compress_file(input_path, temp_path, size=args.size, quality=args.quality)
    if args.in_place:
        temp_path.replace(input_path)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
